### ISSUE TITLE
TCE: Constant pool is in inaccesible address space

### ISSUE THREAD

**Issue by nrother** [pocl/pocl#1149]
The constant pool of LLVM is used to spill large constants to memory. It is placed in address space 0 by default. This is a problem for TCE targets, where AS 0 (which is otherwise only used for the stack) is normally placed in some memory region that cannot be accessed from outside. It would be better if the constant pool would be in AS 2, which is the one used for OpenCL `__constant` memory.

Note that this is not a general compiler problem: For simple TCE applications (without OpenCL) you might only have a single address space, in that case the constant pool must be in AS 0.

Thanks for @TopiLeppanen for pointing this out.

**Comment by TopiLeppanen** [comment]
Yep, can confirm that this is indeed an issue. It can be circumvented by having a 32-bit short immediate, as then the constant pool doesn't seem to be used. Of course if you might rather want to use the long immediate to save on instruction width, you start to bump into this problem.

Possible solutions are being able to either disable the constant pool (then 32-bit long or short immediates would be required) or to being able to change the address space of the constant pool. (In this case maybe to OpenCL constant AS).

**Comment by nrother** [comment]
The problematic code is in [`LLVMTCEBuilder::emitConstantPool`](https://github.com/cpc/openasip/blob/main/openasip/src/applibs/LLVMBackend/LLVMTCEBuilder.cc#L2352): `cpAddrSpaceId` is fixed to 0.

**Comment by pjaaskel** [comment]
Yep, this is a known issue. Last time we looked at it, it was difficult or not possible to relocate the constant pool to a non-default AS, but maybe it has improved in LLVM-side now as it has been years.

**Comment by pjaaskel** [comment]
LLVM/TCE-side issue.
